Kuzzle/Offline/TokenVerifier.cs

```diff
@@ -0,0 +1,129 @@
+    private readonly IQueryReplayer queryReplayer;
+    private readonly IKuzzle kuzzle;
+
+    /// <summary>
+    /// The previous username logged in
+    /// </summary>
+    private string previousKUID = "";
+
+    public TokenVerifier(IOfflineManager offlineManager, IKuzzle kuzzle) {
+      this.offlineManager = offlineManager;
+      this.kuzzle = kuzzle;
+      this.authController = kuzzle.GetAuth();
+
+      queryReplayer = offlineManager.GetQueryReplayer();
+      subscriptionRecoverer = offlineManager.GetSubscriptionRecoverer();
+      kuzzle.GetEventHandler().UserLoggedIn += OnUserLoggedIn;
+    }
+
+    /// <summary>
+    /// Return true if the Token is valid
+    /// </summary>
+    public async Task<bool> IsTokenValid() {
+      JObject response = await authController.CheckTokenAsync(kuzzle.AuthenticationToken);
+
+      if (response == null || response["valid"] == null) return false;
+
+      bool tokenValid = (bool)response["valid"];
+
+
+
+      if (tokenValid
+        && offlineManager.RefreshedTokenDuration > -1
+        && offlineManager.MinTokenDuration > -1) {
+
+        Int64 remainingTime = (Int64) new DateTime(1970, 1, 1)
+          .AddMilliseconds((Int64)response["expiresAt"])
+          .Subtract(DateTime.UtcNow)
+          .TotalMilliseconds;
+
+        if (offlineManager.MinTokenDuration > remainingTime) {
+          try {
+            await authController.RefreshTokenAsync(new TimeSpan(offlineManager.MinTokenDuration * 10000));
+          } catch (Exception) {
+
+          }
+        }
+        return true;
+      }
+
+      return tokenValid;
+    }
+
+    /// <summary>
+    /// This is used to verify if the user that has logged in
+    /// is the same that before, if not this will Reject every query in the Queue
+    /// and clear all subscriptions, otherwise this will replay the Queue if she is waiting.
+    /// </summary>
+    public void OnUserLoggedIn(object sender, UserLoggedInEvent e) {
+
+      if (previousKUID != e.Kuid) {
+
+        if (offlineManager.AutoRecover && queryReplayer.WaitLoginToReplay) {
+            queryReplayer.RejectAllQueries(new UnauthorizeException("Unauthorized", 0));
+            queryReplayer.Lock = false;
+            queryReplayer.WaitLoginToReplay = false;
+        }
+
+        subscriptionRecoverer.Clear();
+      } else {
+
+        if (offlineManager.AutoRecover && queryReplayer.WaitLoginToReplay) {
+            queryReplayer.ReplayQueries();
+            queryReplayer.Lock = false;
+            queryReplayer.WaitLoginToReplay = false;
+        }
+
+        subscriptionRecoverer.RenewSubscriptions();
+      }
+      previousKUID = e.Kuid;
+    }
+
+    /// <summary>
+    /// This will check the token validity,
+    /// and chose what to do before replaying the Queue
+    /// </summary>
+    public async Task CheckTokenToReplay() {
+      if (!offlineManager.AutoRecover) return;
+
+      if (await IsTokenValid()) {
+          queryReplayer.ReplayQueries();
```

Comment:
It seems weird to me that the `TokenVerifier` class has the responsibility of starting replaying the queue.  
IMHO it's more the role of the `OfflineManager` ([here](https://github.com/kuzzleio/sdk-csharp/pull/27/files#diff-e51c4bab1a492bc80dd7bfc44692ee7eR133) )